cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/queue/Job.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright © 2017 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.internal.app.runtime.schedule.queue;
+
+import co.cask.cdap.internal.app.runtime.schedule.ProgramSchedule;
+import co.cask.cdap.proto.Notification;
+import com.google.common.base.Preconditions;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * A scheduled job.
+ */
+public interface Job {
+
+  /**
+   * State of the job. The lifecycle of State is as follows:
+   *
+   *   *-----------------*       *--------------------*         *----------------*
+   *   | PENDING_TRIGGER |  ---> | PENDING_CONSTRAINT |  --->   | PENDING_LAUNCH |
+   *   *-----------------*       *--------------------*         *----------------*
+   */
+  enum State {
+    /**
+     * The Job is awaiting fulfillment of its Trigger.
+     */
+    PENDING_TRIGGER,
+    /**
+     * The Job is awaiting satisfaction of all of its constraints.
+     */
+    PENDING_CONSTRAINT,
+    /**
+     * The Job is awaiting launch.
+     */
+    PENDING_LAUNCH;
+
+    private static final Map<State, State> possibleTransitions = new HashMap<>();
+    static {
+      possibleTransitions.put(PENDING_TRIGGER, PENDING_CONSTRAINT);
+      possibleTransitions.put(PENDING_CONSTRAINT, PENDING_LAUNCH);
+    }
+
+    /**
+     * Checks that the specified State is a valid transition from the current state.
+     */
+    void checkTransition(State nextState) {
+      Preconditions.checkArgument(possibleTransitions.get(this) == nextState,
```

Comment:
Instead of using a static map, it's better to have a field `Set<State> allowedNextStates` and declare the enum as:

```java
  public enum State {

    PENDING_LAUNCH(Collections.<State>emptySet()),
    PENDING_CONSTRAINT(EnumSet.of(PENDING_LAUNCH)),
    PENDING_TRIGGER(EnumSet.of(PENDING_CONSTRAINT));

    private final Set<State> allowedNextState;
    
    State(Set<State> allowedNextState) {
      this.allowedNextState = allowedNextState;
    }

    void checkTransition(State nextState) {
      Preconditions.checkArgument(allowedNextState.contains(nextState));
    }
  }
```